modules/metastorage/src/main/java/org/apache/ignite/metastorage/internal/MetaStorageManager.java

```diff
@@ -75,77 +96,195 @@ public MetaStorageManager(
         });
     }
 
+    /** */
+    public synchronized void deployWatches() {
+        var range = watchAggregator.inferGeneralCriteria();
+        svc.watch(
+            range.getKey(),
+            range.getValue(),
+            watchAggregator.inferGeneralRevision(),
+            watchAggregator.watchListener()).thenAccept(id -> igniteUuid.complete(id)).join();
+        deployed = true;
+    }
+
+    /** */
+    public synchronized CompletableFuture<Long> registerWatch(@Nullable Key key,
+        long revision, @NotNull WatchListener lsnr) {
+        return waitForReDeploy(watchAggregator.add(key, revision, lsnr));
+    }
+
+    /** */
+    public synchronized CompletableFuture<Long> registerWatchByPrefix(@Nullable Key key,
+        long revision, @NotNull WatchListener lsnr) {
+        return waitForReDeploy(watchAggregator.addPrefix(key, revision, lsnr));
+    }
+
+    /** */
+    public synchronized CompletableFuture<Long> registerWatch(@NotNull Collection<Key> keys, long revision,
+        @NotNull WatchListener lsnr) {
+        return waitForReDeploy(watchAggregator.add(keys, revision, lsnr));
+    }
+
+    /** */
+    public synchronized CompletableFuture<Long> registerWatch(@NotNull Key from, @NotNull Key to, long revision,
+        @NotNull WatchListener lsnr) {
+        return waitForReDeploy(watchAggregator.add(from, to, revision, lsnr));
+    }
+
+    /** */
+    public synchronized CompletableFuture<Void> unregisterWatch(long id) {
+        watchAggregator.cancel(id);
+        if (deployed)
+            return updateWatches().thenAccept(v -> {});
+        else
+            return igniteUuid.thenAccept(uuid -> {});
```

Comment:
Why cannot return a completed future, when the global watch have not deployed yet?